created a more thorough overview for devfiles

jc-berger opened the pull request:
For issue [#159](https://github.com/devfile/api/issues/159)

I've applied the comments and feedback from the previous PR, [#25 ]. Due to doc restructure, it was easiest to close out that PR and branch and start anew. 

@elsony and/or @johnmcollier, tagging either of you to see if this new content is thorough and extensive enough for a solid overview on devfiles for users. 

@rkratky, tagging for your awareness as well.

rkratky commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,10 +3,16 @@
 [role="_abstract"]
 
-A devfile describes and defines a cloud native development environment.
+Devfiles are YAML files that describe and define a developer's workspaces or cloud-native-development-environments. Devfiles also describe best practices for end-to-end application development while still being customized or extended to fit a team's current development practices. Instead of forcing a new workflow or tools, devfiles provide instructions for automating what was previously done manu ...
```
```suggestion
Devfiles are YAML files that describe and define developer workspaces or cloud-native development environments. Devfiles also describe best practices for end-to-end application development while providing a way for being customized or extended to fit a team's current development practices. Instead of forcing a new workflow or tools, devfiles provide instructions for automating what was previously done manually.
```

rkratky commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,10 +3,16 @@
+Devfiles are YAML files that describe and define a developer's workspaces or cloud-native-development-environments. Devfiles also describe best practices for end-to-end application development while still being customized or extended to fit a team's current development practices. Instead of forcing a new workflow or tools, devfiles provide instructions for automating what was previously done manu ...
 
-Devfiles are consumed by tools that transform the definition into a cloud workspace composed of multiple cloud native resources. Any number of identical workspaces can be created from the same devfile. They can be shared in various ways including a devfile registry or together with the source code of a project.
+Devfiles are consumed by tools that transform the definition into a cloud workspace composed of multiple containers and cloud-native resources. Also, any number of identical workspaces can be created from the same devfile. This way, workspaces can be worked on by different teams and shared in various ways, including a devfile registry or together with the source code of a project.
```
```suggestion
Devfiles are consumed by tools that transform the definition into a cloud workspace composed of multiple containers and cloud-native resources. Also, any number of identical workspaces can be created from the same devfile. This way, workspaces can be used by different teams and shared in various ways, including a devfile registry or together with the source code of a project.
```

rkratky commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,10 +3,16 @@
+
+* Reducing the gap between development and deployment
+* Finding stacks using the devfile registry
+* Customizing devfiles
```
```suggestion
* Customizing workspaces
```

rkratky reviewed (approved):
Just a few nitpicks.

jc-berger commented:
@rkratky, thanks for the review. I've made a new, squashed commit. 

@elsony and/or @johnmcollier, thorough enough overview do you think?

johnmcollier commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,10 +3,16 @@
+The benefits of devfiles include:
+
+* Reducing the gap between development and deployment
+* Finding stacks using the devfile registry
```
I don't know if I would phrase `Finding stacks using the devfile registry` this way.

While I think a registry of stacks is certainly a benefit to using devfiles, I think it makes more sense to say something like "A registry for finding and hosting devfiles?"

jc-berger replied:
@johnmcollier, thanks. I like your rephrase, but for consistency, I want to have every bullet begin with a gerund. What do you think about:
`Having a registry fro finding and hosting devfiles`

johnmcollier replied:
Yup, I think that  sounds good 👍

jc-berger commented:
@johnmcollier, thanks! New and squashed commit is up. Would you please comment with `LGTM` if you think it's ready to be merged?

@rkratky, once it has the `LGTM` comment, would you please merge, that is, if it looks good to you too!

elsony commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,10 +3,15 @@
+* Producing consistent build and run behaviors
 
 [role="_additional-resources"]
 .Additional resources
```
Add a link to an example of tools that currently have devfile support, e.g. odo and we'll add dev console later after they have adopted it.

elsony commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,10 +3,15 @@
+
+* Reducing the gap between development and deployment
+* Having a registry for finding and hosting devfiles
+* Producing consistent build and run behaviors
```
Add (reword if needed):
Extensible to allow easy devfile customization

elsony reviewed (changes requested):
A couple of things that I'd like to add to the overview:
1. Includes some concrete examples of things that can be included in the devfile, e.g.
   Includes:
   - Guidance about runtime images
   - Example code
   - Build and CI commands
   - Deployment options
2. Add who can use devfiles (stack providers, tools provides, developers)

jc-berger commented:
@elsony, thanks for the review. See the new, squashed commit, addressing your feedback. So, I added context to our `additional resources` so users know why they're going to these links. As you suggested, I added who can use devfiles. Let me know what you think about this new section. 

As far as including example code, commands, guidance on runtimes, and deployment options, were you thinking of taking some content from elsewhere in our docs? Going through the devfile docs we have, I didn't see anything that instantly spoke to any of these topics you suggested. So thoughts on how we'll get this content to users on this overview page?

elsony commented:
> As far as including example code, commands, guidance on runtimes, and deployment options, were you thinking of taking some content from elsewhere in our docs? Going through the devfile docs we have, I didn't see anything that instantly spoke to any of these topics you suggested. So thoughts on how we'll get this content to users on this overview page?

@jc-berger Sorry, maybe I wasn't clear on my initial comment. I only meant to include that list of 4 things (just those 4 entries) as part of the overview docs as the intro.  The user can find the actual details of those from our user guide and schema doc.

elsony commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,28 @@
+* Having a registry for finding and hosting devfiles
+* Producing consistent build and run behaviors
+* Allowing easy devfile customization
+
```
Maybe organize it a bit and put these three consumer usage scenario/benefits under a section, e.g. the section name can be something like "Who would use the devfiles?" or "Consumers of devfiles", etc.

elsony commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,28 @@
+Developers use devfiles to focus only on developing their applications. With devfiles, developers do not need to set up the environment for building and running their applications. Instead, developers have the flexibility to use different tools, like Che and odo.
 
 [role="_additional-resources"]
 .Additional resources
```
Make Additional resources to be a proper section header (e.g. with different font). Currently, it just looks like the rest of the other text.

jc-berger commented:
@elsony, thanks for the feedback. I was on vacation when you did your review, hence, the late reply. The new commit is up.

I like your idea of making the sections of the doc a bigger size, comparable to the title's font size. However, throughout the rest of the doc, each section is simply emboldened, not actually a bigger font size. So to be consistent, I left these section titles bold, not making them bigger. As we continue to improve our docs this year, the doc team can discuss how we want to organize our sections! 

Let me know what you think of the changes and if you think it's ready to merge. Thanks!

elsony reviewed (approved):
/lgtm

ossontoes commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,38 @@
 [role="_abstract"]
 
-A devfile describes and defines a cloud native development environment.
+Devfiles are YAML files that describe and define developer workspaces or cloud-native development environments. Devfiles also describe best practices for end-to-end application development while providing a way for being customized or extended to fit a team's current development practices. Instead of forcing a new workflow or tools, devfiles provide instructions for automating what was previously ...
```
At some RH writing training we were advised to write for "non-native speakers with IELTS level 3" (out of 9). While the meaning of the 2nd sentence seems to be clear, do you think grammar here is on the more advanced side and can cause some readers troubles?

E.g. this part: "while providing a way for being customized or extended to fit". If you decide to rephrase it, possible ways could be "users/you can customize and extend devfiles to..." or "... and can be customized or extended to fit..." (but it's passive voice, it's frowned upon in general).

ossontoes replied:
Also, this part might also be tricky, WDYT? But not too much, it's up to you, of course, if you'd like to keep it.
   "devfiles provide instructions for automating what was previously done manually"
Will this version be a bit more lightweight?
   "for automating manual processes"

ossontoes commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,38 @@
+Devfiles are YAML files that describe and define developer workspaces or cloud-native development environments. Devfiles also describe best practices for end-to-end application development while providing a way for being customized or extended to fit a team's current development practices. Instead of forcing a new workflow or tools, devfiles provide instructions for automating what was previously ...
 
-Devfiles are consumed by tools that transform the definition into a cloud workspace composed of multiple cloud native resources. Any number of identical workspaces can be created from the same devfile. They can be shared in various ways including a devfile registry or together with the source code of a project.
+Devfiles are consumed by tools that transform the definition into a cloud workspace composed of multiple containers and cloud-native resources. Also, any number of identical workspaces can be created from the same devfile. This way, workspaces can be used by different teams and shared in various ways, including a devfile registry or together with the source code of a project.
```
If we intended to reduce passive voice usage, possible rephrasing options would be:
- (development?) tools transform the definition...
- Users can create any number of identical workspaces from the same devfile

But as I wrote earlier I don't believe passive voice is a 100% pure evil, but it's a "no-no" in many RH docs teams.

ossontoes commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,38 @@
+A single devfile can include:
+
+* Guidance about runtime images
+* Examnple code
```
*Example?

ossontoes commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,38 @@
+* Build and CI commands
+* Deployment options
+
+The benefits of devfiles include:
```
If "include" is removed, will the sentence lose some meaning? Can it be seen as noise? Or maybe you wanted some consistency with the previous list (A single devfile can include)?

ossontoes commented on docs/modules/user-guide/partials/con_devfile.adoc:
```diff
@@ -3,15 +3,38 @@
-* xref:devfile-samples.adoc[]
+[role="_additional-resources"]
+
+* To see an example tool that uses devfile support, go to https://developers.redhat.com/products/odo/overview[odo].
+* To learn devfile terminologies, go to xref:api-reference.adoc[]
+* To learn what you can do to a devfile, go to xref:authoring-stacks.adoc[]
+* To update your devfile, go to xref:migrating-to-devfile-v2.adoc[]
+* To learn more about devfiles, go to xref:devfile-samples.adoc[]
```
It's highly project-specific: in several projects I worked with the common word preceding links was "see" (instead of "go to"), but I've never heard it was a rule or company style. Feel free to ignore this if you use "go to" consistently in the docs.

In several trainings and meetings writers discussed xrefs and the general consensus was "Don't use xrefs" :) However, if you never experienced issues while building docs, I don't see why you shouldn't use them. I guess writers ran into troubles while working on some legacy docs.

ossontoes commented:
Hi Jake, thank you for a great and well-written text.
I'm done. My suggestions are mostly about simplifying grammar or reducing noise, but then it's too subjective, feel free to ignore me everywhere :)

jc-berger commented:
@ossontoes, awesome, on point revisions and feedback :) I completely agree about there being too much noise and too complex grammar. 

I've addressed your comments. Please let me know what you think, thanks!

jc-berger commented:
@elsony, I'm thinking about the links in our "additional resources" section. From a doc perspective, links can be troublesome: they can become outdated, encourage users to quickly leave the page, etc. 

I do like directing users' attention after the overview. But instead of giving several links, maybe only one link to the rest our documentation? What do you think?

If you like the idea of cutting down the "additional resources" section, let me know, and I'll get a new commit so we can see how it'd look in the final version.

ossontoes commented:
LGTM, thank you @jc-berger

elsony commented:
@jc-berger How does it look like if there is only one link? Can you provide an example?
